File: packages/engine/src/graph.ts
import { Annotation, END, START, StateGraph } from '@langchain/langgraph';
import type { ActionCategory, AgentAction, Decision, SessionContext } from '@agentgate/shared';
import { classifierNode } from './nodes/classifier.ts';
import { retrieverNode } from './nodes/retriever.ts';
import { judgeNode } from './nodes/judge.ts';
import { decisionGateNode } from './nodes/decisionGate.ts';
import { patternDetectorNode } from './nodes/patternDetector.ts';
import type { GraphState, JudgeVerdict, RetrievedPolicy } from './state.ts';
import { startTrace, type Trace } from './trace.ts';

const last = <T,>(_: T, next: T) => next;

export const StateAnnotation = Annotation.Root({
  action: Annotation<AgentAction>({ reducer: last }),
  context: Annotation<SessionContext>({ reducer: last }),
  category: Annotation<ActionCategory>({ reducer: last, default: () => 'other' }),
  categoryConfidence: Annotation<number>({ reducer: last, default: () => 0 }),
  policies: Annotation<RetrievedPolicy[]>({ reducer: last, default: () => [] }),
  verdict: Annotation<JudgeVerdict>({
    reducer: last,
    default: () => ({ riskScore: 0, reasoning: '' }),
  }),
  decision: Annotation<Decision>({ reducer: last, default: () => 'escalate' }),
  patternNotes: Annotation<string[]>({ reducer: last, default: () => [] }),
  startedAt: Annotation<number>({ reducer: last, default: () => Date.now() }),
  /** Not part of the evaluation — carried so each node can open a span. */
  trace: Annotation<Trace | null>({ reducer: last, default: () => null }),
});

type FullState = GraphState & { trace: Trace | null };

/** Wrap a node so every execution becomes one LangFuse span. */
function traced<S extends FullState>(
  name: string,
  fn: (state: S) => Partial<GraphState> | Promise<Partial<GraphState>>,
  input: (state: S) => unknown,
) {
  return async (state: S) => {
    const span = state.trace?.span(name, input(state));
    const started = Date.now();
    const out = await fn(state);
    span?.end({ ...out, latencyMs: Date.now() - started });
    return out;
  };
}

export function buildGraph() {
  return new StateGraph(StateAnnotation)
    .addNode(
      'classifier',
      traced('classifier.run', classifierNode, (s) => ({
        tool: s.action.toolName,
        args: s.action.toolArgs,
      })),
    )
    .addNode(
      'policy_retriever',
      traced('policy_retriever.search', retrieverNode, (s) => ({ category: s.category })),
    )
    .addNode(
      'risk_judge',
      traced('risk_judge.evaluate', judgeNode, (s) => ({
        category: s.category,
        policies: s.policies.map((p) => p.name),
      })),
    )
    .addNode(
      'decision_gate',
      traced('decision_gate.decide', decisionGateNode, (s) => ({
        riskScore: s.verdict.riskScore,
      })),
    )
    .addNode(
      'pattern_detector',
      traced('pattern_detector.check', patternDetectorNode, (s) => ({
        decision: s.decision,
        sessionId: s.context.sessionId,
      })),
    )
    .addEdge(START, 'classifier')
    .addEdge('classifier', 'policy_retriever')
    .addEdge('policy_retriever', 'risk_judge')
    .addEdge('risk_judge', 'decision_gate')
    .addEdge('decision_gate', 'pattern_detector')
    .addEdge('pattern_detector', END)
    .compile();
}

let compiled: ReturnType<typeof buildGraph> | null = null;
export function getGraph() {
  compiled ??= buildGraph();
  return compiled;
}

export { startTrace };

File: packages/engine/src/agentgate_engine/graph.py
"""The LangGraph state graph.

    classifier → policy_retriever → risk_judge → decision_gate → pattern_detector
"""

from __future__ import annotations

import time
from typing import Any, Awaitable, Callable

from langgraph.graph import END, START, StateGraph

from .nodes.classifier import classifier_node
from .nodes.decision_gate import decision_gate_node
from .nodes.judge import judge_node
from .nodes.pattern_detector import pattern_detector_node
from .nodes.retriever import retriever_node
from .state import GraphState

NodeFn = Callable[[GraphState], Awaitable[dict[str, Any]] | dict[str, Any]]


def traced(name: str, fn: NodeFn, describe: Callable[[GraphState], Any]) -> NodeFn:
    """Wrap a node so every execution becomes one LangFuse span with its latency."""

    async def wrapped(state: GraphState) -> dict[str, Any]:
        span = state["trace"].span(name, describe(state))
        started = time.perf_counter()
        try:
            out = fn(state)
            if hasattr(out, "__await__"):
                out = await out  # type: ignore[misc]
            span.end({"latency_ms": round((time.perf_counter() - started) * 1000)})
            return out  # type: ignore[return-value]
        except Exception as err:  # noqa: BLE001
            span.end(
                {
                    "error": str(err),
                    "latency_ms": round((time.perf_counter() - started) * 1000),
                }
            )
            raise

    return wrapped


def build_graph() -> Any:
    graph = StateGraph(GraphState)

    graph.add_node(
        "classifier",
        traced(
            "classifier.run",
            classifier_node,
            lambda s: {"tool": s["action"].tool_name, "args": s["action"].tool_args},
        ),
    )
    graph.add_node(
        "policy_retriever",
        traced("policy_retriever.search", retriever_node, lambda s: {"category": s["category"]}),
    )
    graph.add_node(
        "risk_judge",
        traced(
            "risk_judge.evaluate",
            judge_node,
            lambda s: {"category": s["category"], "policies": [p.name for p in s["policies"]]},
        ),
    )
    graph.add_node(
        "decision_gate",
        traced(
            "decision_gate.decide",
            decision_gate_node,
            lambda s: {"risk_score": s["verdict"].risk_score},
        ),
    )
    graph.add_node(
        "pattern_detector",
        traced(
            "pattern_detector.check",
            pattern_detector_node,
            lambda s: {"decision": s["decision"], "session_id": s["context"].session_id},
        ),
    )

    graph.add_edge(START, "classifier")
    graph.add_edge("classifier", "policy_retriever")
    graph.add_edge("policy_retriever", "risk_judge")
    graph.add_edge("risk_judge", "decision_gate")
    graph.add_edge("decision_gate", "pattern_detector")
    graph.add_edge("pattern_detector", END)

    return graph.compile()


_compiled: Any | None = None


def get_graph() -> Any:
    global _compiled
    if _compiled is None:
        _compiled = build_graph()
    return _compiled
